state-chain/chains/src/sol/api.rs

```diff
@@ -201,94 +228,80 @@ impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
 	}
 
 	pub fn transfer(
-		transfer_param: Vec<TransferAssetParams<Solana>>,
+		transfer_params: Vec<TransferAssetParams<Solana>>,
 	) -> Result<Vec<Self>, SolanaTransactionBuildingError> {
 		// Lookup the current Aggkey
 		let agg_key = Environment::lookup_account(SolanaEnvAccountLookupKey::AggKey)?;
 		let (nonce_account, durable_nonce) = Environment::nonce_account()?;
 		let compute_price = Environment::compute_price()?;
 
-		// split the transfer params into Native and Token assets
-		let (native, token) = transfer_param
+		let mut sol_environment = HashMap::new();
+		let mut token_environments = HashMap::new();
+		transfer_params
 			.into_iter()
-			.partition::<Vec<TransferAssetParams<Solana>>, _>(|param| param.asset == SolAsset::Sol);
-
-		Ok(vec![
-			// Create native transfer instruction sets
-			native
-				.into_iter()
-				.map(|transfer_param| {
-					SolanaInstructionBuilder::transfer_native(
+			.map(|transfer_param| {
+				let transfer_instruction_set = match transfer_param.asset {
+					SolAsset::Sol => SolanaInstructionBuilder::transfer_native(
 						transfer_param.amount,
 						transfer_param.to,
 						agg_key,
 						nonce_account,
 						compute_price,
-					)
-				})
-				.collect::<Vec<_>>(),
-			if token.is_empty() {
-				vec![]
-			} else {
-				// Only do environment lookup if needed
-				let vault_program =
-					Environment::lookup_account(SolanaEnvAccountLookupKey::VaultProgram)?;
-				let vault_program_data_account = Environment::lookup_account(
-					SolanaEnvAccountLookupKey::VaultProgramDataAccount,
-				)?;
-				let token_vault_pda_account =
-					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenVaultPdaAccount)?;
-				let token_mint_pubkey =
-					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenMintPubkey)?;
-				let token_program_id = SolAddress::from_str(TOKEN_PROGRAM_ID)
-					.expect("Preset Solana Token program ID must be valid.");
-				let token_vault_ata = Environment::lookup_account(
-					SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount,
-				)?;
-				let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
-					.expect("Preset Solana System Program ID account must be valid.");
-
-				// Build the Transfer instruction set for Token transfers.
-				token
-					.into_iter()
-					.map(|transfer_param| {
+					),
+					SolAsset::SolUsdc => {
```

Comment:
Should this not match on any asset? 

Ie:

```suggestion
					token_asset => {
						// ..
						let sol_token_environment = Environment::get_token_environment(
							&mut token_environments,
							token_asset,
						)?;
```

Otherwise we still only support USDC 😅

Reply:
same here, the idea was to have a match arm for each of the future assets we could have. But I can see it would be basically the same code so we can reuse. will change